python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py

```diff
@@ -0,0 +1,1282 @@
+                B_h and B_H : the matrix of L2 scalar product between basis function and fine and coarse snapshot
+        """
+        assert len(self.reducedBasis) !=0, f"need computation of reduced basis"
+
+        interpolateOperator = self.createInterpolator(self.tbCoarse, self.tbFine)
+        InterpCoarseSnaps = []
+        for snap in self.coarseSnapShotList:
+            InterpCoarseSnaps.append(interpolateOperator.interpolate(snap))
+
+        coeffCoarse = np.zeros((self.N, self.N))
+        coeffFine = np.zeros((self.N, self.N))
+
+        for i in range(self.N):
+            for j in range(self.N):
+                coeffCoarse[i,j] = self.l2ScalarProductMatrix.energy(InterpCoarseSnaps[i],self.reducedBasis[j])
+                coeffFine[i,j] = self.l2ScalarProductMatrix.energy(self.fineSnapShotList[i],self.reducedBasis[j])
+
+        return coeffCoarse, coeffFine
+
+    """
+    Check convergence
+    """
+    def checkConvergence(self,Ns=10, Xi_test=None, regulParam=1.e-10):
+        """
+            check convergence of the offline step
+        """
+        assert self.tbCoarse is not None, f"Coarse toolbox needed for computing coarse Snapshot. set doRectification->True"
+
+        if self.worldcomm.isMasterRank():
+            print(f"[NIRB] Compute offline convergence error")
+
+        if Xi_test is None:
+            s = self.Dmu.sampling()
+            s.sampling(Ns,'log-random')
+            vector_mu = s.getVector()
+        else:
+            vector_mu = Xi_test
+
+        Ntest = len(vector_mu)
+
+        interpolationOperator = self.createInterpolator(self.tbCoarse, self.tbFine)
+
+        soltest =[]
+        soltestFine=[]
+        for mu in vector_mu:
+            soltestFine.append(self.getToolboxSolution(self.tbFine, mu))
+            uH = self.getToolboxSolution(self.tbCoarse, mu)
+            soltest.append(interpolationOperator.interpolate(uH))
+
+        tabcoef = np.zeros((Ntest, self.N))
+        tabcoefuh = np.zeros((Ntest, self.N))
+
+        for j in range(Ntest):
+            for k in range(self.N):
+                tabcoef[j,k] = self.l2ScalarProductMatrix.energy(soltest[j],self.reducedBasis[k])
+                tabcoefuh[j,k] = self.l2ScalarProductMatrix.energy(soltestFine[j],self.reducedBasis[k])
+
+        Unirb = self.Xh.element()
+        UnirbRect = self.Xh.element()
+        Uhn = self.Xh.element()
+
+        Error = {'N':[], 'l2(uh-uHn)':[], 'l2(uh-uHn)rec':[], 'l2(uh-uhn)' : [], 'l2(uh-uH)':[]}
+
+        nb = 0
+        pas = 1 if (self.N<50) else 5 
+        for i in tqdm(range(1,self.N+1,pas), desc=f"[NIRB] Compute convergence error:", ascii=False, ncols=100):
+            nb = i
+            # Get rectification matrix
+            BH = self.coeffCoarse[:nb, :nb]
+            Bh = self.coeffFine[:nb, :nb]
+            R = np.zeros((nb,nb))
+            for s in range(nb):
+                R[s,:] = (np.linalg.inv(BH.transpose() @ BH + regulParam*np.eye(nb))@BH.transpose()@Bh[:,s])
```

Comment:
avoid calling `.inv`, rather call `.solve` with multiple right hand sides, @thomas-saigre

Reply:
Yes the correction was made in another branch (in commit 1991f58a5869faee625b4c19f20f5b5bb9ad7394 on branch 2038-nirb-greedy), we may cherry-pick this commit to put the correction in here